Tasks/Common/compression-common/FileExtractor.ts

```diff
@@ -0,0 +1,179 @@
+        return CompressedFile.Tarball;
+    } else if (file.endsWith('.zip')) {
+        return CompressedFile.Zip;
+    } else if (file.endsWith('.7z')) {
+        return CompressedFile.SevenZip;
+    } else {
+        throw new Error(tl.loc('UnsupportedFileExtension'));
+    }
+}
+
+type Extractor = (file: string, destination: string) => Promise<void>;
+
+/** Extract a zip file with the 'unzip' utility. */
+async function unzipExtract(file: string, destination: string): Promise<void> {
+    console.log(tl.loc('UnzipExtractFile', file));
+
+    const unzip: tr.ToolRunner = tl.tool('unzip');
+    unzip.arg(file);
+    await unzip.exec(<tr.IExecOptions>{ cwd: destination });
+}
+
+/** Extract a zip file using PowerShell. */
+async function powershellExtract(file: string, destination: string): Promise<void> {
+    console.log(tl.loc('UnzipExtractFile', file));
+
+    // build the powershell command
+    const escapedFile = file.replace(/'/g, "''").replace(/"|\n|\r/g, ''); // double-up single quotes, remove double quotes and newlines
+    const escapedDest = destination.replace(/'/g, "''").replace(/"|\n|\r/g, '');
+    const command = `
+        $ErrorActionPreference = 'Stop';
+        try
+        {
+            Add-Type -AssemblyName System.IO.Compression.FileSystem
+        }
+        catch { };
+        [System.IO.Compression.ZipFile]::ExtractToDirectory('${escapedFile}', '${escapedDest}')`;
+
+    // change the console output code page to UTF-8
+    const chcpPath = path.join(process.env.windir, "system32", "chcp.com");
+    await tl.exec(chcpPath, '65001');
+
+    // run powershell
+    const powershell: tr.ToolRunner = tl.tool('powershell')
+        .line('-NoLogo -Sta -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command')
+        .arg(command);
+    await powershell.exec();
+}
+
+/** Extract a file using 7zip. */
+const sevenZipExtract = (isWindows: boolean) => async (file: string, destination: string): Promise<void> => {
+    console.log(tl.loc('SevenZipExtractFile', file));
+
+    const sevenZipLocation = isWindows ? path.join(__dirname, '7zip/7z.exe') : tl.which('7z', true);
+    const sevenZip = tl.tool(sevenZipLocation);
+
+    sevenZip.arg('x');
+    sevenZip.arg('-o' + destination);
+    sevenZip.arg(file);
+
+    const execResult = sevenZip.execSync();
+    if (execResult.code !== tl.TaskResult.Succeeded) {
+        tl.debug('execResult: ' + JSON.stringify(execResult));
+    }
+}
+
+/* Extract a compressed tar archive using the 'tar' utiltity. */
+async function tarExtract(file: string, destination: string): Promise<void> {
+    console.log(tl.loc('TarExtractFile', file));
+    const tr: tr.ToolRunner = tl.tool('tar');
+    tr.arg(['xzC', destination, '-f', file]);
+    tr.exec();
+}
+
+/** Choose an extractor function based on the file type and agent's operating system. */
+function pickExtractor(fileType: CompressedFile, hostOS: string): Extractor {
+    const isWindows = hostOS === 'win32';
```

Comment:
You may want to consider using this code: `const isWindows = os.type().match(/^Win/);`  (See: https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/Common/java-common/java-common.ts)

This would keep you from needing to take the parameter for `hostOS`.

Reply:
`os.type()` can "return anything" (there's no predefined set of strings it can return).  `os.platform()` (or equivalently `process.platform`), on the other hand, only returns one of a predefined set of values.  There's no need to regex, just check for `win32`.

Passing the OS type as a parameter makes this code easier to unit test.  If the function is figuring out the OS it is running on by side effect, there would be no way on my Windows dev box for me to test that the right extraction tools are invoked on Linux / macOS (barring some spooky mechanism of intercepting that function call).